designs/vpc_load_balancing.md

```diff
@@ -0,0 +1,84 @@
+# Load Balance VPC Endpoints Across a Set of VPCs
+
+Author: @mjlshen
+
+Last Updated: 02/01/2023
+
+## Summary
+
+`aws-vpce-operator` (AVO)'s `VpcEndpoint` CustomResource (CR) sometimes needs to be used at scale to create VPC
+Endpoints within a set of VPCs.
+
+### Current State
+
+The current state of the VpcEndpoint API requires that consumers either use the cluster's own subnets/VPC or specify a list of subnets that must
+be part of the same VPC. Since AWS has a [quota on the maximum number of VPC Endpoints per VPC](https://docs.aws.amazon.com/vpc/latest/userguide/amazon-vpc-limits.html#vpc-limits-endpoints),
+which starts at 50 and is generally easy to raise to 200, when AVO is heavily used at scale it can run into this limit.
+
+```go
+// Vpc represents the configuration for the AWS VPC to create the VPC Endpoint in
+type Vpc struct {
+    // +kubebuilder:validation:Optional
+
+    // AutoDiscoverSubnets will instruct the controller to use the subnets associated with this ROSA cluster if true.
+    AutoDiscoverSubnets bool `json:"autoDiscoverSubnets,omitempty"`
+
+    // SubnetIds is a list of subnet ids to associate with the VPC Endpoint, which must all be in the same VPC.
+    // If more than one is specified, each subnet must be in a different Availability Zone.
+    // Ref: https://docs.aws.amazon.com/vpc/latest/privatelink/create-interface-endpoint.html
+    SubnetIds []string `json:"subnetIds,omitempty"`
+}
+```
+
+## Motivation
+
+This change is important because:
+
+- It will prevent the need of a higher level orchestrator to stamp out `VpcEndpoint` CRs load balanced across a set of VPCs
+
+## Relevant Stories
+
+- [OSD-13906](https://issues.redhat.com/browse/OSD-13906) Creating VPC Endpoints in the OSD Transit Account
+- [OSD-15012](https://issues.redhat.com/browse/OSD-15012) Load Balancing VPC Endpoints across VPCs
+
+## Goals
+
+- Add an optional set of VPC IDs specified in `.spec.vpc.ids`
+- Pick out private subnet IDs from the provided list of VPCs with the aid of `kubernetes.io/role/internal-elb`
```

Comment:
Are there any alternatives to tagging all subnets AVO should consider with `kubernetes.io/role/internal-elb`?

One I briefly considered was instead of providing a set of VPC IDs, provide a list of....list of subnets that make of VPCs, but that seemed too messy

Reply:
Auto-discovery via tags is a common standard. This is also described in the ALB controller here: https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/deploy/subnet_discovery/